File: CPP/src/core/field/field.cc
# include "field.hh"

Core::Field::Field ( void )
{
}
      
Core::Field::Field ( int x , int y )
{
   for ( int i = 0; i < y; i++ )
   {
      vector< int > row;
      for ( int j = 0; j < x; j++ )
      {
         row.push_back ( 0 );
      }
      contents.push_back ( row );
   }
   
   width = x;
   height = y;
}

Core::Field::~Field ( void )
{
}
      
bool Core::Field::canBeInserted ( Core::Piece piece , int x , int y )
{
   // Check sizes
   if ( x + piece.getWidth () > width )
   {
      return ( false );
   }
   
   if ( y + piece.getHeight () > height )
   {
      return ( false );
   }
   
   // Validate positions
   bool all_ok = true;
   for ( int field_y = y , piece_y = 0; piece_y < piece.getHeight () && all_ok; piece_y++ , field_y++ )
   {
      for ( int field_x = x , piece_x = 0; piece_x < piece.getWidth () && all_ok; piece_x++ , field_x++ )
      {
         if ( contents[ field_y ][ field_x ] > 0 ) // Space used
         {
            if ( piece.getPoint ( piece_x , piece_y ) > 0 )
            {
               all_ok = false; // Collition
            }
         }
      }
   }
   
   return ( all_ok );
}
      
void Core::Field::insertPiece ( Core::Piece piece , int x , int y )
{
   // Copy positions
   for ( int field_y = y , piece_y = 0; piece_y < piece.getHeight (); piece_y++ , field_y++ )
   {
      for ( int field_x = x , piece_x = 0; piece_x < piece.getWidth (); piece_x++ , field_x++ )
      {
         if ( piece.getPoint ( piece_x , piece_y ) > 0 )
         {
            contents[ field_y ][ field_x ] = piece.getPoint ( piece_x , piece_y );
         }
      }
   }
   
   return;
}
      
void Core::Field::removePiece ( int piece_id )
{
   for ( int y = 0; y < height; y++ )
   {
      for ( int x = 0; x < width; x++ )
      {
         if ( contents[ y ][ x ] == piece_id )
         {
            contents[ y ][ x ] = 0;
         }
      }
   }
   
   return;
}

File: Python/core/Field.py
#!/usr/bin/env python3 

from .Matrix import Matrix

class Field ( Matrix ):
   """Field representation"""
   def __init__ ( self , x , y ):
      super ( Field , self ).__init__ ()
      for i in range ( 0 , y ):
         row = []
         for j in range ( 0 , x ):
            row.append ( 0 )
         self._contents.append ( row )
         
      self._width = x
      self._height = y
      
      print ( "Field created with size:", self._width, "x", self._height )
      print ( "Field values stored:", self.getWidth (), "x", self.getHeight () )
      
   def canBeInserted ( self , piece , x , y ):
      if ( x + piece.getWidth () ) > self._width:
         return False
      
      if ( y + piece.getHeight () ) > self._height:
         return False
      
      # Validate positions
      all_ok = True
      field_y = y
      piece_y = 0
      
      while ( piece_y < piece.getHeight () ) and ( all_ok == True ):
         field_x = x
         piece_x = 0
         while ( piece_x < piece.getWidth () ) and ( all_ok == True ):
            if self._contents[ field_y ][ field_x ] > 0: # Space used
               if piece.getPoint ( piece_x , piece_y ) > 0:
                  all_ok = False
            piece_x += 1
            field_x += 1
         piece_y += 1
         field_y += 1
      
      return all_ok
   
   def insertPiece ( self , piece , x , y ):
      print ( "Inserting piece:" )
      piece.print ()
      print ( "In coordinates:" , x , "," , y )
      field_y = y
      for piece_y in range ( 0 , piece.getHeight () ):
         field_x = x
         for piece_x in range ( 0 , piece.getWidth () ):
            print ( "Checking coordinate:",piece_x , ",", piece_y )
            if piece.getPoint ( piece_x , piece_y ) > 0:
               self._contents[ field_y ][ field_x ] = piece.getPoint ( piece_x , piece_y )
            field_x += 1
         field_y += 1
         
   def removePiece ( self , piece_id ):
      for i in range ( 0 , self._height ):
         for j in range ( 0 , self._width ):
            if self._contents[ i ][ j ] == piece_id:
               self._contents[ i ][ j ] = 0
